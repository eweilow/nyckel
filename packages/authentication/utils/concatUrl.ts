import { URL } from "url";

export function concatUrl(origin: string, path: string) {
  const url = new URL(origin);
  url.pathname = path;

  return new URL(url.href).href;
}
