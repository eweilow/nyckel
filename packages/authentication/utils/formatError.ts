export function formatAPIValidationError(error: string) {
  return "[Nyckel] Encountered a validation error: " + error;
}

export function formatAPIError(
  error: string,
  description: string | null,
  endpoint: string
) {
  let msg = `[Nyckel] Encountered an Auth0 API error on '${endpoint}': ${error}`;
  if (description != null) {
    msg += ", " + description;
  }
  return msg;
}
