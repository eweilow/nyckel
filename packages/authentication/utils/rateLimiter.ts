export function createRateLimiter() {
  const store = new Map<
    string,
    {
      now: number;
      limit: number;
      remaining: number;
      used: number;
      resetMillis: number;
      untilNextRefresh: number;
    }
  >();

  const delaySteps = 200;

  let limiter = {
    delay(id: string) {
      if (!store.has(id)) {
        return 0;
      }

      const { now: then, limit, used, untilNextRefresh } = store.get(id)!;

      const correctedUsed = Math.ceil(
        Math.max(0, used - (Date.now() - then) / untilNextRefresh)
      );

      const value = Math.min(
        untilNextRefresh,
        Math.max(0, untilNextRefresh * Math.pow(correctedUsed / limit, 5))
      );

      return delaySteps * Math.floor(value / delaySteps);
    },
    async wait(id: string) {
      const delay = limiter.delay(id);
      if (delay > 0) {
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    },
    update(id: string, limit: number, remaining: number, resetMillis: number) {
      const used = limit - remaining;
      const resetsIn = resetMillis - Date.now();
      const untilNextRefresh = resetsIn / used;

      store.set(id, {
        now: Date.now(),
        limit,
        remaining,
        used,
        resetMillis,
        untilNextRefresh
      });
    },
    updateFromResponse(
      id: string,
      response: { headers: { get: (name: string) => string | null } }
    ) {
      const limit = parseInt(response.headers.get("x-ratelimit-limit")!, 10);
      const remaining = parseInt(
        response.headers.get("x-ratelimit-remaining")!,
        10
      );
      const resetTime =
        parseInt(response.headers.get("x-ratelimit-reset")!, 10) * 1000;

      limiter.update(id, limit, remaining, resetTime);
    }
  };

  return limiter;
}
