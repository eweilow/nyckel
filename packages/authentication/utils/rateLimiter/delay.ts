const delaySteps: number = 200;

export function calculateRateLimiterDelay(
  then: number,
  now: number,
  limit: number,
  used: number,
  untilNextRefresh: number
) {
  const timeSinceThen = now - then;
  const correctedUsed = Math.ceil(
    Math.max(0, used - timeSinceThen / untilNextRefresh)
  );

  const value = Math.min(
    untilNextRefresh,
    Math.max(0, untilNextRefresh * Math.pow(correctedUsed / limit, 5))
  );

  return delaySteps * Math.floor(value / delaySteps);
}
