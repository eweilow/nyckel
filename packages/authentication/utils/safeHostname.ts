import { format } from "url";

// Base implementation from https://github.com/expressjs/express/blob/b8e50568af9c73ef1ade434e92c60d389868361d/lib/request.js
// License:
/*
(The MIT License)

Copyright (c) 2009-2014 TJ Holowaychuk <tj@vision-media.ca>
Copyright (c) 2013-2014 Roman Shtylman <shtylman+expressjs@gmail.com>
Copyright (c) 2014-2015 Douglas Christopher Wilson <doug@somethingdoug.com>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/
function pickFirst(hostHeader?: string): string | null {
  if (!hostHeader || hostHeader == null) {
    return null;
  }
  return hostHeader.split(",")[0].trim();
}

export function getSafeHostname(
  protocol: string,
  xForwardedHostHeader: string | undefined,
  hostHeader: string | undefined,
  isTrusted: (ip: string, val: 0) => boolean
) {
  let host!: string;
  const forwardedHost = pickFirst(xForwardedHostHeader);
  if (forwardedHost == null || !isTrusted(forwardedHost, 0)) {
    host = pickFirst(hostHeader)!;
  }

  return format({
    protocol: protocol,
    host
  });
}
