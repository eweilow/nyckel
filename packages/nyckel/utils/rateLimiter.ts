export function createRateLimiter() {
  const store = new Map<
    string,
    {
      now: number;
      limit: number;
      remaining: number;
      used: number;
      resetMillis: number;
      untilNextRefresh: number;
    }
  >();

  const delaySteps = 200;

  let limiter = {
    delay(id: string) {
      if (!store.has(id)) {
        return 0;
      }

      const { now: then, limit, used, untilNextRefresh } = store.get(id)!;

      const correctedUsed = Math.ceil(
        Math.max(0, used - (Date.now() - then) / untilNextRefresh)
      );

      const value = Math.min(
        untilNextRefresh,
        Math.max(0, untilNextRefresh * Math.pow(correctedUsed / limit, 5))
      );
      return delaySteps * Math.floor(value / delaySteps);
    },
    async wait(id: string) {
      const delay = limiter.delay(id);
      if (delay > 0) {
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    },
    update(id: string, limit: number, remaining: number, resetMillis: number) {
      const used = limit - remaining;
      const resetsIn = resetMillis - Date.now();
      const untilNextRefresh = resetsIn / used;

      store.set(id, {
        now: Date.now(),
        limit,
        remaining,
        used,
        resetMillis,
        untilNextRefresh
      });
    }
  };

  return limiter;
}
