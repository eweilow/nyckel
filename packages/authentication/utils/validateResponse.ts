import { formatAPIError } from "./formatError";

type ErrorResponseTokenBody = {
  error: string;
  error_description: string;
};

export function verifyAPIResponse<T>(
  code: number,
  response: T | ErrorResponseTokenBody,
  endpoint: string
): T {
  if ("error" in response) {
    throw new Error(
      formatAPIError(response.error, response.error_description, endpoint)
    );
  }
  if (code >= 400 && code < 600) {
    throw new Error(formatAPIError("got status " + code, null, endpoint));
  }
  return response;
}
