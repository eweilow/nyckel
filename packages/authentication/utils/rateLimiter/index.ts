import { calculateRateLimiterDelay } from "./delay";

export function createRateLimiter() {
  const store = new Map<
    string,
    {
      now: number;
      limit: number;
      remaining: number;
      used: number;
      resetMillis: number;
      untilNextRefresh: number;
    }
  >();

  let limiter = {
    get store() {
      return store;
    },
    delay(id: string) {
      if (!store.has(id)) {
        return 0;
      }

      const { now: then, limit, used, untilNextRefresh } = store.get(id)!;

      return calculateRateLimiterDelay(
        then,
        Date.now(),
        limit,
        used,
        untilNextRefresh
      );
    },
    async wait(id: string) {
      const delay = limiter.delay(id);
      if (delay > 0) {
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    },
    update(id: string, limit: number, remaining: number, resetMillis: number) {
      const used = limit - remaining;
      const resetsIn = resetMillis - Date.now();
      const untilNextRefresh = resetsIn / used;

      store.set(id, {
        now: Date.now(),
        limit,
        remaining,
        used,
        resetMillis,
        untilNextRefresh
      });
    },
    updateFromResponse(
      id: string,
      response: { headers: { get: (name: string) => string | null } }
    ) {
      const limit = parseInt(response.headers.get("x-ratelimit-limit")!, 10);
      const remaining = parseInt(
        response.headers.get("x-ratelimit-remaining")!,
        10
      );
      const resetTime =
        parseInt(response.headers.get("x-ratelimit-reset")!, 10) * 1000;

      limiter.update(id, limit, remaining, resetTime);
    }
  };

  return limiter;
}
