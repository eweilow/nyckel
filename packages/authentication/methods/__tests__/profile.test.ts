import fetch from "node-fetch";
import { GlobalAuthenticationConfig, getUserInfo } from "../..";
import { createRateLimiter } from "../../utils/rateLimiter";

jest.mock("node-fetch");

jest.mock("../../jwt/decode", () => {
  return {
    decodeJWT: jest.fn(() => ({ sub: "sub" }))
  };
});

jest.mock("../../utils/rateLimiter", () => {
  const instance = {
    rateLimiter: {
      wait: jest.fn(),
      updateFromResponse: jest.fn()
    },
    createRateLimiter: jest.fn(() => {
      return instance.rateLimiter;
    })
  };

  return instance;
});

const config: GlobalAuthenticationConfig = {
  audience: "param:audience",
  authorizationDomain: "param:authorizationDomain",
  clientId: "param:clientId",
  clientSecret: "param:clientSecret",
  urls: {
    authorization: "url:authorization",
    issuer: "url:issuer",
    jwks: "url:jwks",
    logout: "url:logout",
    token: "url:token",
    userinfo: "url:userinfo"
  }
};

const baseData = {
  sub: "sub",
  name: "name",
  nickname: "nickname",
  picture: "picture",
  email: "email",
  email_verified: false,
  updated_at: "2018-10-21"
};

describe("getUserInfo", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should wait for rate limiting correctly", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return baseData;
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    await getUserInfo("accessToken", config);

    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    const rateLimiter = require("../../utils/rateLimiter")
      .rateLimiter as ReturnType<typeof createRateLimiter>;
    expect(rateLimiter.wait).toBeCalledWith("sub");
    expect(rateLimiter.wait).toBeCalledTimes(1);
  });

  it("should update rate limiting correctly", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return baseData;
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    await getUserInfo("accessToken", config);

    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    const rateLimiter = require("../../utils/rateLimiter")
      .rateLimiter as ReturnType<typeof createRateLimiter>;

    expect(rateLimiter.updateFromResponse).toBeCalledWith("sub", fetchResponse);
    expect(rateLimiter.updateFromResponse).toBeCalledTimes(1);
  });

  it("should re-attempt if rate limited", async () => {
    let responseInstance1 = {
      status: 429,
      json: jest.fn(() => {
        throw new Error("Rate limited!");
      })
    };
    let responseInstance2 = {
      status: 200,
      json: jest.fn(() => {
        return baseData;
      })
    };
    ((fetch as any) as jest.Mock)
      .mockReturnValueOnce(Promise.resolve(responseInstance1))
      .mockReturnValueOnce(Promise.resolve(responseInstance2));

    await getUserInfo("accessToken", config);

    const rateLimiter = require("../../utils/rateLimiter")
      .rateLimiter as ReturnType<typeof createRateLimiter>;

    expect(rateLimiter.updateFromResponse).toHaveBeenNthCalledWith(
      1,
      "sub",
      responseInstance1
    );
    expect(rateLimiter.updateFromResponse).toHaveBeenNthCalledWith(
      2,
      "sub",
      responseInstance2
    );
    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    expect(fetch).toBeCalledTimes(2);
    expect(rateLimiter.wait).toBeCalledTimes(2);
    expect(rateLimiter.updateFromResponse).toBeCalledTimes(2);
  });

  it("should fail if error", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return {
          error: "oh no",
          error_description: "something bad happened"
        };
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    const promise = getUserInfo("accessToken", config);
    await expect(promise).rejects.toThrow();
    const err = await promise.catch(err => err);

    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    expect(fetch).toBeCalledTimes(1);
    expect(err.message).toMatchSnapshot();
  });

  describe("fail on missing data", () => {
    for (const key of Object.keys(baseData)) {
      it(`throws if ${key} is missing from body`, async () => {
        const fetchResponse = {
          status: 200,
          json: jest.fn(() => {
            return { ...baseData, [key]: undefined };
          })
        };
        ((fetch as any) as jest.Mock).mockReturnValueOnce(
          Promise.resolve(fetchResponse)
        );
        const promise = getUserInfo("accessToken", config);
        await expect(promise).rejects.toThrowErrorMatchingSnapshot();
      });
    }

    for (const key of Object.keys(baseData)) {
      it(`throws if ${key} is an invalid type`, async () => {
        const fetchResponse = {
          status: 200,
          json: jest.fn(() => {
            return { ...baseData, [key]: 3 };
          })
        };
        ((fetch as any) as jest.Mock).mockReturnValueOnce(
          Promise.resolve(fetchResponse)
        );
        const promise = getUserInfo("accessToken", config);
        await expect(promise).rejects.toThrowErrorMatchingSnapshot();
      });
    }
  });

  it("should return correctly", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return baseData;
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    const data = await getUserInfo("accessToken", config);
    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    expect(data).toMatchSnapshot();

    expect(data.updatedAt).toBeInstanceOf(Date);
  });

  it("logs to console.info", async () => {
    const original = process.env.NODE_ENV;
    process.env.NODE_ENV = "development";

    const spy = jest.spyOn(console, "info").mockImplementation(() => {});

    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return baseData;
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    await getUserInfo("accessToken", config);

    expect(spy.mock.calls).toMatchSnapshot();
    spy.mockRestore();
    process.env.NODE_ENV = original;
  });
});
