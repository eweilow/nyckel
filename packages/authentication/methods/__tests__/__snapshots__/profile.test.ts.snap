// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserInfo fail on missing data throws if email is an invalid type 1`] = `"Expected decoded.email to be type string, but got number"`;

exports[`getUserInfo fail on missing data throws if email is missing from body 1`] = `"Expected decoded userinfo to have property 'email'"`;

exports[`getUserInfo fail on missing data throws if email_verified is an invalid type 1`] = `"Expected decoded.email_verified to be type boolean, but got number"`;

exports[`getUserInfo fail on missing data throws if email_verified is missing from body 1`] = `"Expected decoded userinfo to have property 'email_verified'"`;

exports[`getUserInfo fail on missing data throws if name is an invalid type 1`] = `"Expected decoded.name to be type string, but got number"`;

exports[`getUserInfo fail on missing data throws if name is missing from body 1`] = `"Expected decoded userinfo to have property 'name'"`;

exports[`getUserInfo fail on missing data throws if nickname is an invalid type 1`] = `"Expected decoded.nickname to be type string, but got number"`;

exports[`getUserInfo fail on missing data throws if nickname is missing from body 1`] = `"Expected decoded userinfo to have property 'nickname'"`;

exports[`getUserInfo fail on missing data throws if picture is an invalid type 1`] = `"Expected decoded.picture to be type string, but got number"`;

exports[`getUserInfo fail on missing data throws if picture is missing from body 1`] = `"Expected decoded userinfo to have property 'picture'"`;

exports[`getUserInfo fail on missing data throws if sub is an invalid type 1`] = `"Expected decoded.sub to be type string, but got number"`;

exports[`getUserInfo fail on missing data throws if sub is missing from body 1`] = `"Expected decoded userinfo to have property 'sub'"`;

exports[`getUserInfo fail on missing data throws if updated_at is an invalid type 1`] = `"Expected decoded.updated_at to be type string, but got number"`;

exports[`getUserInfo fail on missing data throws if updated_at is missing from body 1`] = `"Expected decoded userinfo to have property 'updated_at'"`;

exports[`getUserInfo logs to console.info 1`] = `
Array [
  Array [
    "[Nyckel] requesting userinfo at url:userinfo",
  ],
]
`;

exports[`getUserInfo should fail if error 1`] = `
Array [
  Array [
    "url:userinfo",
    Object {
      "headers": Object {
        "Authorization": "Bearer accessToken",
      },
    },
  ],
]
`;

exports[`getUserInfo should fail if error 2`] = `"[Nyckel] Encountered an Auth0 API error on 'url:userinfo': oh no, something bad happened"`;

exports[`getUserInfo should re-attempt if rate limited 1`] = `
Array [
  Array [
    "url:userinfo",
    Object {
      "headers": Object {
        "Authorization": "Bearer accessToken",
      },
    },
  ],
  Array [
    "url:userinfo",
    Object {
      "headers": Object {
        "Authorization": "Bearer accessToken",
      },
    },
  ],
]
`;

exports[`getUserInfo should return correctly 1`] = `
Array [
  Array [
    "url:userinfo",
    Object {
      "headers": Object {
        "Authorization": "Bearer accessToken",
      },
    },
  ],
]
`;

exports[`getUserInfo should return correctly 2`] = `
Object {
  "email": "email",
  "emailVerified": false,
  "locale": undefined,
  "name": "name",
  "nickname": "nickname",
  "picture": "picture",
  "sub": "sub",
  "updatedAt": 2018-10-21T00:00:00.000Z,
}
`;

exports[`getUserInfo should update rate limiting correctly 1`] = `
Array [
  Array [
    "url:userinfo",
    Object {
      "headers": Object {
        "Authorization": "Bearer accessToken",
      },
    },
  ],
]
`;

exports[`getUserInfo should wait for rate limiting correctly 1`] = `
Array [
  Array [
    "url:userinfo",
    Object {
      "headers": Object {
        "Authorization": "Bearer accessToken",
      },
    },
  ],
]
`;
